osu.Game.Tests/Visual/Multiplayer/TestSceneMultiplayerSpectator.cs

```diff
@@ -0,0 +1,386 @@
+            {
+                Beatmap.Value = beatmapManager.GetWorkingBeatmap(importedBeatmap);
+                Ruleset.Value = importedBeatmap.Ruleset;
+
+                LoadScreen(spectatorScreen = new MultiSpectatorScreen(playingUserIds.ToArray()));
+            });
+
+            AddUntilStep("wait for screen load", () => spectatorScreen.LoadState == LoadState.Loaded && (!waitForPlayerLoad || spectatorScreen.AllPlayersLoaded));
+        }
+
+        private void start(int userId, int? beatmapId = null) => start(new[] { userId }, beatmapId);
+
+        private void start(int[] userIds, int? beatmapId = null)
+        {
+            AddStep("start play", () =>
+            {
+                foreach (int id in userIds)
+                {
+                    Client.CurrentMatchPlayingUserIds.Add(id);
+                    streamingClient.StartPlay(id, beatmapId ?? importedBeatmapId);
+                    playingUserIds.Add(id);
+                    nextFrame[id] = 0;
+                }
+            });
+        }
+
+        private void finish(int userId, int? beatmapId = null)
+        {
+            AddStep("end play", () =>
+            {
+                streamingClient.EndPlay(userId, beatmapId ?? importedBeatmapId);
+                playingUserIds.Remove(userId);
+                nextFrame.Remove(userId);
+            });
+        }
+
+        private void sendFrames(int userId, int count = 10) => sendFrames(new[] { userId }, count);
+
+        private void sendFrames(int[] userIds, int count = 10)
+        {
+            AddStep("send frames", () =>
+            {
+                foreach (int id in userIds)
+                {
+                    streamingClient.SendFrames(id, nextFrame[id], count);
+                    nextFrame[id] += count;
+                }
+            });
+        }
+
+        private void checkPaused(int userId, bool state)
+            => AddUntilStep($"{userId} is {(state ? "paused" : "playing")}", () => getPlayer(userId).ChildrenOfType<GameplayClockContainer>().First().GameplayClock.IsRunning != state);
+
+        private void checkPausedInstant(int userId, bool state)
+            => AddAssert($"{userId} is {(state ? "paused" : "playing")}", () => getPlayer(userId).ChildrenOfType<GameplayClockContainer>().First().GameplayClock.IsRunning != state);
+
+        private void assertVolume(int userId, double volume)
+            => AddAssert($"{userId} volume is {volume}", () => getInstance(userId).Volume.Value == volume);
+
+        private void waitForCatchup(int userId)
+            => AddUntilStep($"{userId} not catching up", () => !getInstance(userId).GameplayClock.IsCatchingUp);
+
+        private Player getPlayer(int userId) => getInstance(userId).ChildrenOfType<Player>().Single();
+
+        private PlayerArea getInstance(int userId) => spectatorScreen.ChildrenOfType<PlayerArea>().Single(p => p.UserId == userId);
+
+        public class TestSpectatorStreamingClient : SpectatorStreamingClient
```

Comment:
This is effectively a duplicate of the one in a few other test scenes. It's my first priority to clean up after this is merged in - the implementation is slightly different in every case.

Please ignore for now, unless you really want this fixed here.

Reply:
This has now been removed, and all implementations combined as of https://github.com/ppy/osu/pull/12538/commits/21fc0ba43baea6e0cb115952ab21df06f1d1c003